### ISSUE TITLE
Training on custom dataset

### ISSUE THREAD

**Issue by furkanpala** [ZeningLin/ViBERTgrid-PyTorch#9]
Hi, thanks for your work and contribution. I have studied on Chargrid and especially BERTgrid. Now, I am trying to understand the ViBERTgrid paper. Could you explain how can I use your implementation on my own custom dataset? Are train scripts generalizeable to custom datasets?
Thanks.

**Comment by ZeningLin** [comment]
Hi Furkan!
Thanks for your interest in this repo. To use a custom dataset, you should know what we need during ViBERTgrid-net's forward and backward propagation:

1.  To construct a ViBERTgrid and apply early-fusion/late-fusion, the **OCR result of a document** is needed. I used the OCR results provided in the label of SROIE/EPHOIE and **stored them in .csv format** (If needed, you may extract OCR results of your dataset through open-source OCR engines like Paddle-OCR or MMOCR). The csv file should contain the following columns: _left, top, right, bot, text, class, pos_neg_, where the first four are the coordinates of the left-top and right-bot vertices, 'class' stands for the key-information categories, 'pos_neg' are categories mentioned in section 3.3 of the paper (_The first one is used to classify each pixel into one of three categories: (1) Not located in any word box; (2) Inside a word box which belongs to any pre-defined field; (3) Inside other word boxes._). 
2. To extract visual information, the **original document images** should be provided. 
3. To supervise the training process, two kinds of labels should be generated. The first one is the **class label** which has a shape of `(N, H, W)`, N is the number of key information categories, H, W are height and width of the corresponding original image. Each pixel is filled with its category one-hot encoding, similar to those of Chargrid's. The second one is the **pos_neg label** in shape (3, H, W). 3 stands for the three classes mentioned in 1), others are the same as the class labels.

To sum up, you should provide the **original document images**, **OCR results of each document in csv format** , **class labels** and **pos_neg labels** before loading your dataset for training. The structure of the raw dataset folder shold be like

```
------raw dataset folder-------
 [dir]-> image
 [dir]-> _label_csv
 [dir]-> _class
 [dir]-> _pos_neg
 train.txt
 test.txt
```

You need to write your own script to generate the above things. You may refer to function `single_label_parser()` in `/pipe/ephoie_data_preprocessing.py` for detailed implementation.  Don't forget to generate train.txt and test.txt which contains filenames for train-set and test-set (Use `random` or `np.random` to make it).

To load your custom dataset, just follow(copy) the method in `data/EPHOIE_dataset.py` and write you own `CUSTOM_dataset.py`. It will automatically generate the batches required during training (including corpus after BERT tokenizing, coordinate tensors, masks for BERT model, etc.). The train scripts can be generalized to all custom datasets which follows the pipeline mentioned above.

**Comment by furkanpala** [comment]
Hi again, I have some questions about the implementation, especially related to BERTgrid. Is there a way that we can discuss, like a mailing platform etc. that you can share with me?

**Comment by ZeningLin** [comment]
> 

email: zening.lin@outlook.com
